File: Blind150/7.Tree/Python/btree_from_inorder_and_preorder.py
# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/description

from typing import List,Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:

        if not preorder or not inorder:
            return None

        root = TreeNode(preorder[0])

        # root_pos_inorder = 0

        # for pos, val in enumerate(inorder):

        #     if val == root.val:

        #         root_pos_inorder = pos

        root_pos_inorder = inorder.index(root.val)

        inorder_left, inorder_right = inorder[:root_pos_inorder], inorder[root_pos_inorder+1:]

        preorder_left, preorder_right =  preorder[1:len(inorder_left)+1], preorder[len(inorder_left)+1:]


        root.left = self.buildTree(preorder_left, inorder_left)
        root.right = self.buildTree(preorder_right, inorder_right)

        return root

File: Blind150/7.Tree/Go/btree_from_inorder_and_preorder.go
// https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/description

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {

	if len(preorder) == 0 || len(inorder) == 0 {
		return nil
	}

	root := &TreeNode{Val: preorder[0]}

	mid := 0

	for pos, val := range inorder {

		if val == root.Val {
			mid = pos
			break
		}
	}

	// Split inorder and preorder lists
	inorderLeft, inorderRight := inorder[:mid], inorder[mid+1:]
	preorderLeft, preorderRight := preorder[1:len(inorderLeft)+1], preorder[len(inorderLeft)+1:]

	root.Left = buildTree(preorderLeft, inorderLeft)
	root.Right = buildTree(preorderRight, inorderRight)

	return root

}
